presto-main/src/main/java/io/prestosql/connector/system/TableCommentSystemTable.java

```diff
@@ -0,0 +1,124 @@
+import static io.prestosql.metadata.MetadataListing.listTables;
+import static io.prestosql.metadata.MetadataUtil.TableMetadataBuilder.tableMetadataBuilder;
+import static io.prestosql.spi.connector.SystemTable.Distribution.SINGLE_COORDINATOR;
+import static io.prestosql.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.util.Objects.requireNonNull;
+
+public class TableCommentSystemTable
+        implements SystemTable
+{
+    private static final SchemaTableName COMMENT_TABLE_NAME = new SchemaTableName("metadata", "table_comments");
+
+    private static final ConnectorTableMetadata COMMENT_TABLE = tableMetadataBuilder(COMMENT_TABLE_NAME)
+            .column("catalog_name", createUnboundedVarcharType())
+            .column("schema_name", createUnboundedVarcharType())
+            .column("table_name", createUnboundedVarcharType())
+            .column("comment", createUnboundedVarcharType())
+            .build();
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+
+    @Inject
+    public TableCommentSystemTable(Metadata metadata, AccessControl accessControl)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return COMMENT_TABLE;
+    }
+
+    @Override
+    public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession connectorSession, TupleDomain<Integer> constraint)
+    {
+        Optional<String> catalogFilter = stringFilter(constraint, 0);
+        Optional<String> schemaFilter = stringFilter(constraint, 1);
+        Optional<String> tableFilter = stringFilter(constraint, 2);
+
+        Session session = toSession(transactionHandle, connectorSession);
+        Builder table = InMemoryRecordSet.builder(COMMENT_TABLE);
+
+        for (String catalog : filter(listCatalogs(session, metadata, accessControl).keySet(), catalogFilter)) {
+            QualifiedTablePrefix prefix = FilterUtil.tablePrefix(catalog, schemaFilter, tableFilter);
+
+            Set<SchemaTableName> names = new HashSet<>();
+            try {
+                names = listTables(session, metadata, accessControl, prefix);
+            }
+            catch (PrestoException ignored) {
+                // listTables throws an exception if cannot connect the database
+            }
+
+            for (SchemaTableName name : names) {
+                QualifiedObjectName tableName = new QualifiedObjectName(prefix.getCatalogName(), name.getSchemaName(), name.getTableName());
+                Optional<String> comment = Optional.empty();
+                try {
+                    Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
```

Comment:
I think this can simplify to
```java
comment = metadata.getTableHandle(session, tableName)
        .map(handle -> metadata.getTableMetadata(session, handle))
        .map(metadata -> metadata.getMetadata().getComment());